daemon/system-info.sh

```diff
@@ -393,6 +393,49 @@ elif pgrep "kubelet"; then
   HOST_IS_K8S_NODE="true"
 fi
 
+# ------------------------------------------------------------------------------------------------
+# Detect instance metadata for VMs running on cloud providers
+
+CLOUD_TYPE="unknown"
+CLOUD_INSTANCE_TYPE="unknown"
+CLOUD_INSTANCE_REGION="unknown"
+
+if [ "${VIRTUALIZATION}" != "none" ] && command -v curl > /dev/null 2>&1; then
+  # Try AWS IMDSv2
+  if [ "${CLOUD_TYPE}" = "unknown" ]; then
+    AWS_IMDS_TOKEN="$(curl --fail -s -m 5 --noproxy "*" -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")"
+    if [ -n "${AWS_IMDS_TOKEN}" ]; then
+      CLOUD_TYPE="AWS"
+      CLOUD_INSTANCE_TYPE="$(curl --fail -s -m 5 --noproxy "*" -H "X-aws-ec2-metadata-token: $TOKEN" -v "http://169.254.169.254/latest/meta-data/instance-type")"
+      CLOUD_INSTANCE_REGION="$(curl --fail -s -m 5 --noproxy "*" -H "X-aws-ec2-metadata-token: $TOKEN" -v "http://169.254.169.254/latest/meta-data/placement/region")"
+    fi
+  fi
+
+  # Try GCE computeMetadata v1
+  if [ "${CLOUD_TYPE}" = "unknown" ]; then
+    if [ -n "$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1")" ]; then
+      CLOUD_TYPE="GCP"
+      CLOUD_INSTANCE_TYPE="$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/machine-type")"
+      [ -n "$CLOUD_INSTANCE_TYPE" ] && CLOUD_INSTANCE_TYPE=$(basename "$CLOUD_INSTANCE_TYPE")
+      CLOUD_INSTANCE_REGION="$(curl --fail -s -m 5 --noproxy "*" -H "Metadata-Flavor: Google" "http://metadata.google.internal/computeMetadata/v1/instance/zone")"
+      [ -n "$CLOUD_INSTANCE_REGION" ] && CLOUD_INSTANCE_REGION=$(basename "$CLOUD_INSTANCE_REGION") && CLOUD_INSTANCE_REGION=${CLOUD_INSTANCE_REGION%-*}
+    fi
+  fi
+
+  # Try Azure IMDS
+  if [ "${CLOUD_TYPE}" = "unknown" ]; then
+    AZURE_IMDS_DATA="$(curl --fail -s -m 5 -H "Metadata: true" --noproxy "*" "http://169.254.169.254/metadata/instance?version=2021-10-01")"
+    if [ -n "${AZURE_IMDS_DATA}" ]; then
+      CLOUD_TYPE="Azure"
+      CLOUD_INSTANCE_TYPE="$(curl --fail -s -m 5 -H "Metadata: true" --noproxy "*" "http://169.254.169.254/metadata/instance/compute/vmSize?version=2021-10-01&format=text")"
+      CLOUD_INSTANCE_REGION="$(curl --fail -s -m 5 -H "Metadata: true" --noproxy "*" "http://169.254.169.254/metadata/instance/compute/location?version=2021-10-01&format=text")"
+    fi
+  fi
+fi
```

Comment:
I think we either need to test it or have commented out until we can test it. Otherwise, we can get garbage values.

Reply:
> @Ferroin after last commits, I still have everything set as `unknown` on `Arch VM`:
> 
> ```shell
> NETDATA_CONTAINER_OS_NAME=none
> NETDATA_CONTAINER_OS_ID=none
> NETDATA_CONTAINER_OS_ID_LIKE=none
> NETDATA_CONTAINER_OS_VERSION=none
> NETDATA_CONTAINER_OS_VERSION_ID=none
> NETDATA_CONTAINER_OS_DETECTION=none
> NETDATA_CONTAINER_IS_OFFICIAL_IMAGE=false
> NETDATA_HOST_OS_NAME=Arch Linux
> NETDATA_HOST_OS_ID=arch
> NETDATA_HOST_OS_ID_LIKE=unknown
> NETDATA_HOST_OS_VERSION=unknown
> NETDATA_HOST_OS_VERSION_ID=unknown
> NETDATA_HOST_OS_DETECTION=/etc/os-release
> NETDATA_HOST_IS_K8S_NODE=false
> NETDATA_SYSTEM_KERNEL_NAME=Linux
> NETDATA_SYSTEM_KERNEL_VERSION=5.16.15-arch1-1
> NETDATA_SYSTEM_ARCHITECTURE=x86_64
> NETDATA_SYSTEM_VIRTUALIZATION=oracle
> NETDATA_SYSTEM_VIRT_DETECTION=systemd-detect-virt
> NETDATA_SYSTEM_CONTAINER=unknown
> NETDATA_SYSTEM_CONTAINER_DETECTION=systemd-detect-virt
> NETDATA_SYSTEM_CPU_LOGICAL_CPU_COUNT=2
> NETDATA_SYSTEM_CPU_VENDOR=GenuineIntel
> NETDATA_SYSTEM_CPU_MODEL=Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
> NETDATA_SYSTEM_CPU_FREQ=2700000000
> NETDATA_SYSTEM_CPU_DETECTION=lscpu
> NETDATA_SYSTEM_TOTAL_RAM=1017008128
> NETDATA_SYSTEM_RAM_DETECTION=procfs
> NETDATA_SYSTEM_TOTAL_DISK_SIZE=21474836480
> NETDATA_SYSTEM_DISK_DETECTION=sysfs
> NETDATA_CLOUD_TYPE=unknown
> NETDATA_CLOUD_INSTANCE_TYPE=unknown
> NETDATA_CLOUD_INSTANCE_REGION=unknown
> ```

This is expected behavior if you are either not running in a supported cloud environment, or don’t have `curl` installed.

Reply:
In theory we could test this in a GHA runner (they're just on-demand Azure VMs), though I’m not sure if they actually have access to the required link-local address.